serializers/avro/src/main/java/io/pravega/schemaregistry/serializer/avro/impl/AvroGenericDeserializer.java

```diff
@@ -10,39 +10,41 @@
 package io.pravega.schemaregistry.serializer.avro.impl;
 
 import com.google.common.base.Preconditions;
-import io.pravega.schemaregistry.serializer.avro.schemas.AvroSchema;
 import io.pravega.schemaregistry.client.SchemaRegistryClient;
 import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.serializer.avro.schemas.AvroSchema;
 import io.pravega.schemaregistry.serializer.shared.impl.AbstractDeserializer;
 import io.pravega.schemaregistry.serializer.shared.impl.EncodingCache;
 import io.pravega.schemaregistry.serializer.shared.impl.SerializerConfig;
 import org.apache.avro.Schema;
 import org.apache.avro.generic.GenericDatumReader;
 import org.apache.avro.io.BinaryDecoder;
 import org.apache.avro.io.DecoderFactory;
-
+import org.apache.commons.lang3.tuple.Pair;
 import javax.annotation.Nullable;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.concurrent.ConcurrentHashMap;
 
 public class AvroGenericDeserializer extends AbstractDeserializer<Object> {
-    private final ConcurrentHashMap<SchemaInfo, Schema> knownSchemas;
+    private final ConcurrentHashMap<Pair<SchemaInfo, SchemaInfo>, GenericDatumReader<Object>> knownSchemaReaders;
```

Comment:
we dont need a pair here. the readerschema is supplied by the user at the time of construction of the deserializer in the factory. 
so the reader schema is fixed (or null). only the writer schema is variable based on the schema that was used for writing. so we can simply keep it as a map of writerSchema to generic reader.

in case of null, the call into this method actually passes writer schema as reader schema itself.

Reply:
In `io.pravega.schemaregistry.serializers.SerializerTest > testMultiformatDeserializers` test when calling AvroGenericDeserializer constructor - readerSchema is null, but when calling deserialize method readerSchemaInfo is not null, so we can't use reader schema from constructor, because  when calling deserialize method, another readerSchema can be provided.
If changing AvroGenericDeserializer similar to AvroDeserializer using reader from constructor and writerSchemaData as key to datumReadersMap then  `testMultiformatDeserializers ` is failing:
```Java
io.pravega.schemaregistry.serializers.SerializerTest > testMultiformatDeserializers FAILED
    java.lang.NullPointerException
        at org.apache.avro.Schema.getAliases(Schema.java:1868)
        at org.apache.avro.Schema.applyAliases(Schema.java:1796)
        at org.apache.avro.generic.GenericDatumReader.getResolver(GenericDatumReader.java:128)
        at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:142)
        at io.pravega.schemaregistry.serializer.avro.impl.AvroGenericDeserializer.deserialize(AvroGenericDeserializer.java:57)
        at io.pravega.schemaregistry.serializers.MultipleFormatDeserializer.deserialize(MultipleFormatDeserializer.java:43)
        at io.pravega.schemaregistry.serializer.shared.impl.AbstractDeserializer.deserialize(AbstractDeserializer.java:123)
        at io.pravega.schemaregistry.serializers.SerializerTest.testMultiformatDeserializers(SerializerTest.java:85)
```